numba/jitclass/base.py

```diff
@@ -0,0 +1,415 @@
+    registered_typingctx = set()
+
+    def __init__(self, class_type, methods, typingctx, targetctx):
+        self.class_type = class_type
+        self.methods = methods
+        self.typingctx = typingctx
+        self.targetctx = targetctx
+        self.register_once_per_typingctx()
+        self.register_once_per_class()
+
+    def register_once_per_typingctx(self):
+        """
+        Register information to the typing context.
+        The registration is executed once for typing context.
+        """
+        if self.typingctx not in self.registered_typingctx:
+            self.implement_attribute_typing(self.typingctx)
+            self.registered_typingctx.add(self.typingctx)
+
+    def register_once_per_class(self):
+        """
+        Register code-generation information to the target context.
+        The registration is executed once per (class, targetcontext)
+        """
+        registry = imputils.Registry()
+        if (type(self), self.targetctx) not in self.registered_targetctx:
+            self.implement_constructor(registry)
+            self.implement_attribute(registry)
+            self.targetctx.insert_func_defn(registry.functions)
+            self.targetctx.insert_attr_defn(registry.attributes)
+            self.registered_targetctx[type(self), self.targetctx] = registry
+
+    def register(self):
+        """
+        Register to the frontend and backend.
+        """
+        class_type = self.class_type
+        instance_type = class_type.instance_type
+        self.implement_frontend(instance_type)
+        self.implement_backend(instance_type)
+
+    def implement_frontend(self, instance_type):
+        pass
+
+    @classmethod
+    def implement_attribute_typing(cls, typingctx):
+        typingctx.insert_attributes(ClassAttribute(typingctx))
+
+    def implement_backend(self, instance_type):
+        registry = imputils.Registry()
+        self.register_methods(registry, instance_type)
+        self.targetctx.insert_func_defn(registry.functions)
+        self.targetctx.insert_attr_defn(registry.attributes)
+
+    @classmethod
+    def implement_constructor(cls, registry):
+        registry.register(ctor_impl)
+
+    def register_methods(self, registry, instance_type):
+        # Add methods
+        for meth in instance_type.jitmethods:
+            self.implement_method(registry, instance_type, meth)
+
+    @classmethod
+    def implement_attribute(cls, registry):
+        registry.register_attr(attr_impl)
+
+    def implement_method(self, registry, instance_type, attr):
+        # TODO: we should be able to refactor this so that the registration
+        #       can be at the top-level instead of at per instance type.
+        @registry.register
+        @imputils.implement((instance_type, attr), types.VarArg(types.Any))
+        def imp(context, builder, sig, args):
+            method = instance_type.jitmethods[attr]
+            method.compile(sig)
+            cres = method._compileinfos[sig.args]
```

Comment:
Rather than accessing `_compileinfos`, you could stick to non-internal APIs, e.g.:
```
call = context.get_function(types.Dispatcher(method), sig)
out = call(builder, args)
```